Fix DTensor get_mesh_from_args when first arg is not a tensor

Aminsed opened the pull request:
Fixes #167915

## Summary
`OpSchema.get_mesh_from_args()` only inspected the first argument to find the device mesh. If the first arg wasn't a DTensor (e.g., a string or scalar for custom ops), it would fail with:
```
ValueError: Cannot find device mesh from args for op : <op_name>
```

This is a problem for custom ops where the first argument may be a non-tensor type (e.g., a name string), but subsequent arguments are DTensors.

## Root Cause
The code only checked `self.args_schema[0]`:
```python
first_arg = self.args_schema[0]
if isinstance(first_arg, (DTensorSpec, OpStrategy)):
    mesh = first_arg.mesh
# ...
else:
    raise ValueError(...)
```

## Fix
Update `get_mesh_from_args()` to scan all positional arguments to find the first `DTensorSpec` or `OpStrategy`, rather than only checking the first argument.

## Test Plan
Added `test_register_sharding_non_tensor_first_arg` which:
1. Defines a custom op with a string as the first arg and a tensor as the second
2. Registers a sharding rule for it
3. Verifies the op works correctly with DTensor inputs

cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx @H-Huang @awgu @fegin @fduwjj @wconstab @msaroufim @dcci @ezyang

Aminsed committed:
Fix DTensor get_mesh_from_args when first arg is not a tensor

OpSchema.get_mesh_from_args() only inspected the first argument to find
the device mesh. If the first arg wasn't a DTensor (e.g., a string or
scalar), it would fail with 'Cannot find device mesh from args'.

This fix updates get_mesh_from_args() to scan all positional arguments
to find the first DTensorSpec or OpStrategy, rather than only checking
the first argument.

Fixes #167915

Aminsed commented:
@pytorchbot label "module: dtensor" "oncall: distributed"

ali-m07 commented:
## Review

### ✅ Positive
- Good fix for DTensor get_mesh_from_args
- Important bug fix for tensor operations

### 💡 Suggestions
- Add tests to prevent regression
- Consider edge cases
- Document the fix if it changes behavior

### ✅ Approval
Important bug fix. Looks good!

ezyang reviewed (commented):
It seems plausible but I worry about list of optional tensor cannot check first elem only

Aminsed committed:
Address reviewer feedback

1. Fix List[Optional[Tensor]] handling in get_mesh_from_args()

   - If the first element was None (common in Optional tensor lists),
     the mesh would not be found from subsequent elements
   - Now scans ALL elements in the list/tuple to find the first valid
     DTensorSpec or OpStrategy

2. Add proper test cleanup for custom ops

   - Added tearDown() method to TestRegisterSharding class
   - Cleans up custom ops registered under 'test_dtensor::' namespace

Aminsed commented:
@wconstab @stmcgovern I checked [PR #173489](https://github.com/pytorch/pytorch/pull/173489). It fixes kwargs-based mesh extraction (e.g., eye.m_out). This PR addresses a different path: OpSchema.get_mesh_from_args() in sharding propagation still inspects only args_schema[0], so cases where the first arg is non-tensor and a later arg is DTensor can still fail. So they’re related.

wconstab commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -448,18 +448,23 @@ def get_mesh_from_args(self, validate: bool = True) -> DeviceMesh:
-        else:
+        mesh = None
+        # Scan all args to find the first DTensorSpec/OpStrategy (not just the first arg)
+        for arg in self.args_schema:
```
ok found it- try_find_mesh_from_args in _utils.py - can you see if you can just call that from here instead?

ezyang reviewed (approved):
This seems straightforwardly beneficial and not much incremental complexity on top.

ezyang commented:
@pytorchbot merge
